# Creation of the Dataframe

In [1]:
#import libraries
!pip install beautifulsoup4
import requests
import bs4 as BeautifulSoup
import pandas as pd

In [2]:
#url of wikipedia
url="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [3]:
s=requests.Session()
response=s.get(url,timeout=10)
response

<Response [200]>

In [4]:
soup=BeautifulSoup.BeautifulSoup(response.text,'html.parser')
torontotable=soup.find('table',{'class':"wikitable sortable"})

In [5]:
df=pd.read_html(str(torontotable))
# convert list to dataframe
df=pd.DataFrame(df[0])
print(df.head())

  Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront


Clean the data

In [6]:
# Ignore cells with a borough that is Not assigned. Remove rows
df2=df[~df["Borough"].str.contains("Not assigned")==True]

In [7]:
#M5A is already combined in one row with the neighborhoods separated with a comma as shown in row 4  in the below table. (used BautifulSoup)
df2.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [8]:
#If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough.
#Check For a Substring in the Column "Neighbourhood"
NeighbourhoodNA=df2.loc[df2['Neighbourhood'].str.contains("Not assigned",case=False)]
NeighbourhoodNA.head()

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

There are no cells with Not assigned Neighborhood.

In [9]:
df2.shape

(103, 3)

The dataframe has 103 rows and 3 columns

# Get the latitude and the longitude coordinates of each neighborhood

In [10]:
!pip install geocoder
import geocoder # import geocoder

     |████████████████████████████████| 98 kB 7.3 MB/s  eta 0:00:01


Import the .csv file with coordinates.
To difficult to construct the loop function

def lat_lng_coords

#### initialize your variable to None
lat_lng_coords = None

##### loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

In [11]:

import types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share the notebook.
client_04abeae85fc94048947b8ed7bfe090d5 = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='glH5Z5O9um1CkoL5IbdSTzORAzFbWCuEGFWpZRoseoST',
    ibm_auth_endpoint="https://iam.cloud.ibm.com/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3.eu-geo.objectstorage.service.networklayer.com')

body = client_04abeae85fc94048947b8ed7bfe090d5.get_object(Bucket='capstoneproject-donotdelete-pr-egev5iahubzuav',Key='Geospatial_Coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1.head()


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Create the dataframe

In [12]:
FinalTable=pd.merge(df2,df_data_1, on='Postal Code')
FinalTable.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

# Explore and cluster the neighborhoods in Toronto

Import libraries

In [13]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [14]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(FinalTable['Borough'].unique()),
        FinalTable.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


Use geopy library to get the latitude and longitude values of Toronto city.

In [15]:
address = 'Toronto Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.¶

In [19]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(FinalTable['Latitude'], FinalTable['Longitude'], FinalTable['Borough'], FinalTable['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Let's simplify the above map and segment and cluster only the neighborhoods in Downtown Toronto. So let's slice the original dataframe and create a new dataframe of the Downtown Toronto data.

In [21]:
DowntownToronto_data = FinalTable[FinalTable['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
DowntownToronto_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Let's get the geographical coordinates of Downtown Toronto.

In [22]:
address = 'Downtown Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6541737, -79.38081164513409.


#### As with all of Toronto , let's visualize Downtown Toronto the neighborhoods in it.

In [23]:
# create map of Downtown Toronto using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(DowntownToronto_data['Latitude'], DowntownToronto_data['Longitude'], DowntownToronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

Next, I am going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [86]:

CLIENT_ID = 'AF3NMAQEM3DHBZQICGWEJVIO0T1NAOCZSHGD3NMMAS1NO0CM'
CLIENT_SECRET = '4GPVXUK0EGHNE5GABYKY3V3PML5EWL4G0INMT0R0XB0RTT3J'
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: AF3NMAQEM3DHBZQICGWEJVIO0T1NAOCZSHGD3NMMAS1NO0CM
CLIENT_SECRET:4GPVXUK0EGHNE5GABYKY3V3PML5EWL4G0INMT0R0XB0RTT3J


#### Let's explore the first neighborhood in our dataframe

Get the neighborhood's name

In [26]:
DowntownToronto_data.loc[0, 'Neighbourhood']

'Regent Park, Harbourfront'

Get the neighborhood's latitude and longitude values.

In [28]:
neighborhood_latitude = DowntownToronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = DowntownToronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = DowntownToronto_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


#### let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters

First, let's create the GET request URL. Name your URL url.

In [29]:
LIMIT=100
radius=500
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=AF3NMAQEM3DHBZQICGWEJVIO0T1NAOCZSHGD3NMMAS1NO0CM&client_secret=4GPVXUK0EGHNE5GABYKY3V3PML5EWL4G0INMT0R0XB0RTT3J&ll=43.6541737,-79.38081164513409&v=20180605&radius=500&limit=100'

Send the GET request and examine the resutls

In [30]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6044a7eae377ec72d271c80b'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Bay Street Corridor',
  'headerFullLocation': 'Bay Street Corridor, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 95,
  'suggestedBounds': {'ne': {'lat': 43.6586737045, 'lng': -79.37460365419369},
   'sw': {'lat': 43.6496736955, 'lng': -79.38701963607448}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '57eda381498ebe0e6ef40972',
       'name': 'UNIQLO ユニクロ',
       'location': {'address': '220 Yonge St',
        'crossStreet': 'at Dundas St W',
        'lat': 43.65591027779457,
        'lng': -79.38064099181345,
        'labeledLatLngs': [{

From the Foursquare lab in the previous module, I know that all the information is in the items key. Before I proceed, let's borrow the get_category_type function from the Foursquare lab.

In [31]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now I am ready to clean the json and structure it into a pandas dataframe.

In [32]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name      categories        lat        lng
0                       UNIQLO ユニクロ  Clothing Store  43.655910 -79.380641
1  Elgin And Winter Garden Theatres         Theater  43.653394 -79.378507
2                Ed Mirvish Theatre         Theater  43.655102 -79.379768
3                            Indigo       Bookstore  43.653515 -79.380696
4                              LUSH  Cosmetics Shop  43.653557 -79.380400

And how many venues were returned by Foursquare?

In [33]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

95 venues were returned by Foursquare.


 # Explore Neighborhoods in Downtown Toronto


Let's create a function to repeat the same process to all the neighborhoods in Downtown Toronto

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and create a new dataframe called _downtownvenues

In [35]:
downtown_venues = getNearbyVenues(names=DowntownToronto_data['Neighbourhood'],
                                   latitudes=DowntownToronto_data['Latitude'],
                                   longitudes=DowntownToronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


#### Let's check the size of the resulting dataframe

In [36]:
print(downtown_venues.shape)
downtown_venues.head()



(1223, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

#### Let's check how many venues were returned for each neighborhood

In [38]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     64   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              63   
Queen's Park, Ontario Provincial Government                            30   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
St. James Town                                                         81   
St. James Town, Cabbagetown                                            45   
Stn A PO Boxes                                                         99   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         32   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      64   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               63   
Queen's Park, Ontario Provincial Government                             30   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                                94   
Rosedale                                                                 4   
St. James Town                                                          81   
St. James Town, Cabbagetown                                             45   
Stn A PO Boxes                                                          99   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          32   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            59              59   
CN Tower, King and Spadina, Railway Lands, Harb...     15              15   
Central Bay Street                                     64              64   
Christie                                               16              16   
Church and Wellesley                                   75              75   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues

In [39]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 211 uniques categories.


# Analyze Each Neighborhood

In [42]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Lounge  Airport Service  \
0            0        0                   0               0                0   
1            0        0                   0               0                0   
2            0        0                   0               0                0   
3            0        0                   0               0                0   
4            0        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bistro  Boat or Ferry  \
0                   0                         0       0              0   
1                   0                         0       0              0   
2                   0                         0       0              0   
3                   0                         0       0              0   
4                   0                         0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Butcher  Café  Candy Store  \
0                0             0              0        0     0            0   
1                0             0              0        0     0            0   
2                0             0              0        0     0            0   
3                0             0              0        0     0            0   
4                0             0              0        0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0  

In [43]:
downtown_onehot.shape

(1223, 211)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [44]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.066667   
2                                  Central Bay Street     0.015625  0.000000   
3                                            Christie     0.000000  0.000000   
4                                Church and Wellesley     0.026667  0.000000   
5                      Commerce Court, Victoria Hotel     0.000000  0.000000   
6              First Canadian Place, Underground city     0.000000  0.000000   
7                            Garden District, Ryerson     0.000000  0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
10        Queen's Park, Ontario Provincial Government     0.033333  0.000000   
11                          Regent Park, Harbourfront     0.021739  0.000000   
12                           Richmond, Adelaide, King     0.000000  0.000000   
13                                           Rosedale     0.000000  0.000000   
14                                     St. James Town     0.000000  0.000000   
15                        St. James Town, Cabbagetown     0.022222  0.000000   
16                                     Stn A PO Boxes     0.010101  0.000000   
17           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
18                     University of Toronto, Harbord     0.031250  0.000000   

    Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0             0.000000        0.000000              0.0          0.000000   
1             0.066667        0.133333              0.2          0.066667   
2             0.000000        0.000000              0.0          0.000000   
3             0.000000        0.000000              0.0          0.000000   
4             0.000000        0.000000              0.0          0.000000   
5             0.000000        0.000000              0.0          0.000000   
6             0.000000        0.000000              0.0          0.000000   
7             0.000000        0.000000              0.0          0.000000   
8             0.000000        0.000000              0.0          0.000000   
9             0.000000        0.000000              0.0          0.000000   
10            0.000000        0.000000              0.0          0.000000   
11            0.000000        0.000000              0.0          0.000000   
12            0.000000        0.000000              0.0          0.000000   
13            0.000000        0.000000              0.0          0.000000   
14            0.000000        0.000000              0.0          0.000000   
15            0.000000        0.000000              0.0          0.000000   
16            0.000000        0.000000              0.0          0.000000   
17            0.000000        0.000000              0.0          0.000000   
18            0.000000        0.000000              0.0          0.000000   

    American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0              0.000000      0.000000      0.00     0.016949    0.000000   
1              0.000000      0.000000      0.00     0.000000    0.000000   
2              0.000000      0.000000      0.00     0.000000    0.015625   
3              0.000000      0.000000      0.00     0.000000    0.000000   
4              0.013333      0.000000      0.00     0.000000    0.000000   
5              0.040000      0.000000      0.00     0.010000    0.000000   
6              0.030000      0.000000      0.00     0.010000    0.000000   
7              0.000000      0.000000      0.00     0.010000    0.000000   
8              0.000000      0.000000      0.05     0.010000    0.000000   
9              0.000000      0.000000      0.00     0.000000    0.000000   
10             0.000000      0.000000      0.00     0.000000   

In [45]:
downtown_grouped.shape

(19, 211)

#### Let's print each neighborhood along with the top 5 most common venues

In [46]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.08
1    Cocktail Bar  0.05
2          Bakery  0.05
3        Pharmacy  0.03
4  Farmers Market  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.20
1      Airport Lounge  0.13
2       Boat or Ferry  0.07
3  Airport Food Court  0.07
4    Airport Terminal  0.07


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1      Sandwich Place  0.06
2                Café  0.05
3  Italian Restaurant  0.05
4     Bubble Tea Shop  0.03


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3     Baby Store  0.06
4     Restaurant  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1  Japanese Restaurant  0.07
2     Sushi Restaurant  0.07
3           Restaurant  0.04
4              Gay Ba

#### Let's put that into a pandas datafram

First, let's write a function to sort the venues in descending order.

In [47]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [51]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
downtown_venues_sorted = pd.DataFrame(columns=columns)
downtown_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    downtown_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

downtown_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery              Pharmacy   
1        Airport Lounge         Boat or Ferry    Airport Food Court   
2        Sandwich Place                  Café    Italian Restaurant   
3                  Café                  Park            Baby Store   
4   Japanese Restaurant      Sushi Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0        Farmers Market              Beer Bar        Seafood Restaurant   
1      Airport Terminal       Harbor / Marina                  Boutique   
2       Bubble Tea Shop          Burger Joint       Japanese Restaurant   
3            Restaurant           Candy Store        Athletics & Sports   
4               Gay Bar           Yoga Studio  Mediterranean Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop            Restaurant          Shopping Mall  
1               Airport                 Plane                    Bar  
2           Salad Place       Thai Restaurant  Portuguese Restaurant  
3             Nightclub           Coffee Shop     Italian Restaurant  
4                   Pub                  Café            Men's Store

# Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [66]:
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 4, 0, 2, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [71]:
downtown_venues_sorted.drop(['Cluster Labels'], axis=1, inplace=True) #I can't insert 'Cluster Labels' column into the table as it already exists there.
#I should drop the column first and then re run the code

# add clustering labels
downtown_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = DowntownToronto_data

# merge downtown_grouped with downtown_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(downtown_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

downtown_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0                  Café   
4  43.644771 -79.373306               0           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0                  Park                     Bakery               Theater   
1      Sushi Restaurant                      Diner           Yoga Studio   
2           Coffee Shop  Middle Eastern Restaurant        Cosmetics Shop   
3           Coffee Shop               Cocktail Bar             Gastropub   
4          Cocktail Bar                     Bakery              Pharmacy   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Café                   Pub        Breakfast Spot   
1   Fried Chicken Joint                  Café    College Auditorium   
2                  Café   Japanese Restaurant    Italian Restaurant   
3   American Restaurant                   Gym                 Hotel   
4        Farmers Market              Beer Bar    Seafood Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Yoga Studio        Farmers Market         Chocolate Shop  
1     College Cafeteria              Creperie    Distribution Center  
2       Bubble Tea Shop                 Hotel       Ramen Restaurant  
3        Lingerie Store   Moroccan Restaurant         Farmers Market  
4           Cheese Shop            Restaurant          Shopping Mall

Finally, let's visualize the resulting clusters

In [73]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighbourhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine Clusters

examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, then assign a name to each cluster.

#### Cluster 1

In [74]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0                  Café   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0                  Café   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0                  Café   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue      3rd Most Common Venue  \
0                   Park                     Bakery   
1       Sushi Restaurant                      Diner   
2            Coffee Shop  Middle Eastern Restaurant   
3            Coffee Shop               Cocktail Bar   
4           Cocktail Bar                     Bakery   
5         Sandwich Place                       Café   
7                   Café                 Restaurant   
8               Aquarium                      Hotel   
9                  Hotel                       Café   
10            Restaurant                       Café   
12           Coffee Shop      Vietnamese Restaurant   
15    Seafood Restaurant               Cocktail Bar   
16           Coffee Shop         Italian Restaurant   
17                  Café                      Hotel   
18   Japanese Restaurant           Sushi Restaurant   

            4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                         Theater                  Café                   Pub   
1                     Yoga Studio   Fried Chicken Joint                  Café   
2                  Cosmetics Shop                  Café   Japanese Restaurant   
3                       Gastropub   American Restaurant                   Gym   
4                        Pharmacy        Farmers Market              Beer Bar   
5              Italian Restaurant       Bubble Tea Shop          Burger Joint   
7                 Thai Restaurant        Clothing Store                   Gym   
8                            Café    Italian Restaurant   Fried Chicken Joint   
9                      Restaurant   Japanese Restaurant           Salad Place   
10                          Hotel                   Gym   American Restaurant   
12  Vegetarian / Vegan Restaurant    Mexican Restaurant                   Bar   
15             Italian Restaurant   Japanese Restaurant              Beer Bar   
16             Chinese Restaurant            Restaurant                   Pub   
17                     Restaurant   Japanese Restaurant                   Gym   
18                     Restaurant               Gay Bar           Yoga Studio   

       7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0             Breakfast Spot           Yoga Studio        Farmers Market   
1         College Auditorium     College Cafeteria              Creperie   
2         Italian Restaurant       Bubble Tea Shop                 Hotel   
3                      Hotel        Lingerie Store   Moroccan Restaurant   
4         Seafood Restaurant           Cheese Shop            Restaurant   
5        Japanese Restaurant           Salad Place       Thai Restaurant   
7              Deli / Bodega                 Hotel                Bakery   
8                    Brewery        Scenic Lookout            Restaurant   
9         Seafood Restaurant   American Restaurant                Bakery   
10        Italian Restaurant   Japanese Restaurant         Deli / Bodega   
12    

#### Cluster 2

In [75]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14            Playground                 Trail   Monument / Landmark   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14                Market   Martial Arts School        Massage Studio   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14  Mediterranean Restaurant           Men's Store     Mexican Restaurant

#### Cluster 3

In [76]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6            Restaurant           Candy Store    Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6             Nightclub           Coffee Shop     Italian Restaurant

#### Cluster 4

In [77]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
11  Downtown Toronto               3                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11                Bakery   Japanese Restaurant             Bookstore   

   5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
11                   Bar           Yoga Studio  College Arts Building   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11            Beer Store        Sandwich Place             Restaurant

#### Cluster 5

In [78]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               4       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13        Airport Lounge         Boat or Ferry    Airport Food Court   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13      Airport Terminal       Harbor / Marina              Boutique   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13               Airport                 Plane                    Bar

Conclusion: Can't get a meaningful analysis.